homeassistant/components/freebox/switch.py

```diff
@@ -18,49 +20,67 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+@dataclass
+class FreeboxSwitchRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    state_fn: Callable[[FreeboxRouter], Any]
+    on_off_fn: Callable[[FreeboxRouter], Any]
+
+
+@dataclass
+class FreeboxSwitchEntityDescription(
+    SwitchEntityDescription, FreeboxSwitchRequiredKeysMixin
+):
+    """Describes Freebox switch entity."""
+
+
+SWITCH_DESCRIPTIONS = [
+    FreeboxSwitchEntityDescription(
+        key="wifi",
+        name="Freebox WiFi",
+        entity_category=EntityCategory.CONFIG,
+        state_fn=lambda router: router.wifi.get_global_config,
+        on_off_fn=lambda router: router.wifi.set_global_config,
+    )
```

Comment:
I think that `state_fn` and `on_off_fn` are not useful inside the SwitchEntityDescription. They add confusion and since there is only one EntityDescription currently I would keep them inside the `FreeboxSwitch` for now.
```suggestion
    SwitchEntityDescription(
        key="wifi",
        name="Freebox WiFi",
        entity_category=EntityCategory.CONFIG,
    )
```

Reply:
I've seen multiple integrations using this.
OK sure, for one switch it may be overkill, but the plan it to add more in the near future and use this rework PR to clear the platform.

Reply:
Then maybe you should make it more generic, with `state_key` and `on_off_data_fn`?
```suggestion
    FreeboxSwitchEntityDescription(
        key="wifi",
        name="Freebox WiFi",
        entity_category=EntityCategory.CONFIG,
        state_fn=lambda router: router.wifi.get_global_config(),
        state_key="enabled",
        on_off_fn=lambda router, data: router.wifi.set_global_config(data),
        on_off_data_fn=lambda enabled: {"enabled": enabled},
    )
```

```python
@dataclass
class FreeboxSwitchRequiredKeysMixin:
    """Mixin for required keys."""

    state_fn: Callable[[FreeboxRouter], Coroutine[Any, Any, dict[str, Any]]]
    state_key: str
    on_off_fn: Callable[[FreeboxRouter, dict[str, Any]], Coroutine[Any, Any, None]]
    on_off_data_fn: Callable[[bool], dict[str, Any]]
```

```python
    async def _async_set_state(self, enabled: bool):
        """Turn the switch on or off."""
        try:
            on_off_data = self.entity_description.on_off_data_fn(enabled)
            await self.entity_description.on_off_fn(self._router, on_off_data)
        except InsufficientPermissionsError:
            _LOGGER.warning(
                "Home Assistant does not have permissions to modify the Freebox settings. Please refer to documentation"
            )
```

```python
    async def async_update(self) -> None:
        """Get the state and update it."""
        datas = await self.entity_description.state_fn(self._router)
        self._attr_is_on = bool(datas[self.entity_description.state_key])
```

Reply:
I was thinking today and yes, we should keep the description simple for now and will "generify" when needed.

Thanks for the help !